DOC: Reorganization request for Development Workflow docs

rnelsonchem opened the pull request:
As a new contributor, I tried to follow the Development Workflow documentation; however, I found this page to be a little confusing in places. I've reorganized things in a way that seems to make more sense to me now that I've made a code submission via pull request. I'm not a git wizard, so I would appreciate any feedback in case I messed things up. Below are the major changes and remaining questions that I'd like to address. I'm happy to roll a lot of this back, but this addresses some issues I had as a newcomer. 

Major changes:
- Move the rebasing section out of the basic workflow. The first sentence in this section advises not to do this, so it is unclear why this would be a basic workflow issue. Is there another reason that this is important? Perhaps to make sure there are no merge conflicts, but if so, it is not clear to a beginner.
- Also separate out the problems from the rebasing section. It seems like this covers (or could cover) more than just rebasing, so it made sense to make it a separate section.
- Removed core dev notes from the basic workflow. I don't imagine that the core devs need to run through the basic workflow checklist all that often, and these are not commands that a beginner needs to worry about. (Or they probably can't do without commit rights...?)
- Moved some of the discussion of git push from the New Branch section to the end of the Editing section. It just seemed more natural that you would need to worry about pushing to a remote repo after you've done all of your changes.
- Renamed a few links and removed a broken one (Matthew Brett's NiPy comparison link)

Questions:
- The Asking for a Merge section suggests both code reviews via the mailing list and pull requests. Is there a reason to prefer one over the other? Should pull requests only be filed for opened issues? It might be nice to include that.
- I tried to build the documentation in a fresh conda Python 3.4 environment with the minimal dependencies. However, it choked right away because it could not find the plot_directive extension. Of course, installing Matplotlib into this environment fixed the error. However, there is nothing about this dependency in doc/HOWTO_BUILD_DOCS.rst.txt. I'm happy to add something to this document, but I'm not sure that you want to recommend MPL as a Numpy dependency. Is there an alternate method to finding this extension that should be included in the docs?

rnelsonchem committed:
DOC: Update Development Workflow Page

The "Basic Workflow" section of the Development Workflow page was reorganized
to make it clearer to new users and remove/update old material.

Major changes:

- Moved the core developer notes. Core dev notes are confusing in basic usage
  information. Updated a couple of the commands

- Added more commands and reorgaized the "New Feature Branch" section. Took
  some text from "Rebasing" and moved it here, because it was redundant.

- Made many changes to the "Editing Workflow" to clarify commands. Moved some
  git push stuff into this section as well.

- Moved pull request section to just after editing workflow. Removed outdated
  images and simplified the text. Make code review requirements explicit.

- Moved rebasing and troubleshooting sections to the end. Could add more here
  later.

- Links renamed and broken link removed.

rgommers commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -16,94 +16,81 @@ Basic workflow
+3. When finished:
 
    - *Contributors*: push your feature branch to your own Github repo, and
      :ref:`ask for code review or make a pull request <asking-for-merging>`.
```
Maybe update this as well. We don't really want people to ask for code review on their Github repo, we only want PRs.

rnelsonchem replied:
For a code review, should the recommendation be to just email a description and link to the PR to the mailing list?

rgommers replied:
Indeed. But I commented on that below already, so here maybe just remove the "ask for code review"?

rgommers commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -16,94 +16,81 @@ Basic workflow
+   From now on git_ will know that ``my-new-feature`` is related to the
+   ``my-new-feature`` branch in your own github_ repo. This applies to the
+   ``master`` branch as well. You will have to use ``--set-upstream`` for each
+   new branch that you create.
```
Isn't this fairly pointless overhead?

rgommers commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -145,18 +132,23 @@ In more detail
+      git push origin my-new-feature
+
+It may be the case that while you were working on your edits, new commits have
+been added to ``origin`` that affect your work. In this case, follow the
```
Not ``origin`` but ``upstream``.

rgommers commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -195,27 +187,78 @@ Standard acronyms to start the commit message with are::
+=======================================================
+
+When you feel your work is finished, you can file a pull request and
+submit an optional code review. Code reviews are typically submitted when an
```
What does "and submit an optional code review" mean? Maybe ask for a code review on the mailing list?

rgommers replied:
Ah yes, read the below. I'd say something like "When you change an existing API or add a new function to the the API, please send a summary of your change with motivation to the numpy-discussion mailing list - that's the place where decisions on API changes are made."

rgommers commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -195,27 +187,78 @@ Standard acronyms to start the commit message with are::
+
+Pushing changes to the main repo
+================================
+
```
Maybe add a note here that this is only relevant if you have commit rights to Numpy itself?

rgommers commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -195,27 +187,78 @@ Standard acronyms to start the commit message with are::
+
+3. Push to upstream::
+
+        git push upstream my-feature-branch:master
```
This is not quite right. For case (a), a few commits, we still want a merge commit. Reasons: we want to always group related commits, and we don't want to try any but the last commit when using ``git bisect``.

charris replied:
That's right, we've pretty much settled on always making merge commits. When we started on github we tried to keep the history as one linear line, but eventually figured out that that was an abuse of git and that merges made it easier to track and deal with multiple commits, and even for single commits it allows determining who did the merge, as opposed to who made the modifications.

juliantaylor replied:
this will keep the merge commit, there is no rebasing involved.

Though I do not like this section much, we should almost never do this, the two only exceptions are release preparations and backports that already went through a pull request.

I personally also do not like to have a remote setup to point to the numpy repo. This allows for accidental pushes to the main repo.
I usually only have my fork as origin and for pushing to the main repo directly I explicitly have to type the git@url line instead of a remote shorthand. That way the chance of mistakingly pushing to the wrong remote are almost zero.

tacaswell replied:
A tactic I use with mpl is I have a read-only remote called 'upstream' and a read/write remote called 'HOLYCOMMITRIGHTSBATMAN' or some similarly silly name in all capitals so dealing with back-porting isn't too annoying, but chances of accidentally pushing to the wrong remote are low.

rgommers replied:
this will keep the merge commit, there is no rebasing involved.

1a still says rebase then push, which is what I meant.

rgommers commented:
Overall this looks good to me. Thanks @rnelsonchem.

rgommers commented:
I realize that I commented on a few things that were already there and you just reshuffled. Hope that's OK - this seems like a good time to fix those things.

rnelsonchem commented:
@rgommers Thank you so much for the feedback.

You are indeed correct that some of my changes involved moving text that was already present; however, it is good to fix those problems as well while I'm at it.

A question: Regarding your comment on rebasing/merging for the main repo -- Should the two cases be: 1) A few unrelated commits (rebase)  and 2) Any number of related commits (merge)? These section is a bit outside the scope of my git-fu... I pretty much cut-and-pasted this section verbatim from earlier in the document.

A couple of comments:

1) I corrected the code review references. There is now only a very small paragraph at the end of the merge section that recommends this process for API/function changes.

2) I moved and updated the note on using the ``--set-upstream`` flag for ``git push``. That flag isn't necessary for the master branch, but it saves a little typing for new development branches. I moved this info to the editing section, where it is more relevant, and added another comment.

3) I added a note about changes to the main Numpy repo requiring the proper authorization.

rgommers commented:
@rnelsonchem those are indeed the two cases

rgommers commented:
The updates in the last commit all look good to me.

charris commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -117,7 +84,7 @@ Overview
    # hack hack
    git add my_new_file
-   git commit -am 'ENH: some message'
+   git commit -a
```
It is also good practice to do a `git status` before making commits to see what is going to happen, sometimes a `git diff` also. Both can help avoid mistakes.

rgommers replied:
``git status`` before commit makes sense here, I disagree with removing ``-a`` - it's what you want 9 out of 10 times.  ``git add`` is still needed for a new file (which is rare), but not for changes to existing files.

matthew-brett replied:
I agree with Chuck that you should not use `-a` after `git add` though - that's a very bad habit to get into.  I've often carefully done `git add` to select the changes I want and then thoughtlessly done `git commit -a` therefore wiping out all my work on selecting changes.

rnelsonchem commented:
Thank you everyone for your comments. I added some more info about git pull vs git fetch and merge. I also removed the "-a" flag from git commit, and added some explanatory text that describes this as a optional use case. 

At this point, I think this looks pretty good, so I squashed everything into one commit. I'm happy to make more changes, though, if needed.

charris commented on doc/source/dev/gitwash/development_workflow.rst:
```diff
@@ -16,93 +16,63 @@ Basic workflow
-In git >= 1.7 you can ensure that the link is correctly set by using the
-``--set-upstream`` option::
+   However, never use ``git pull`` without indicating the source and
+   destination, as the inherent ambiguity can cause problems. This avoids a
```
The pair `upstream master` is the source, the destination is the current branch. If you want to be consistent with the fetch/merge version `git pull --ff-only upstream master`.

rnelsonchem replied:
@charris -- Just to be clear... Above this there is a merge command that uses "upstream/master" and here there is "upstream master". Are both of these branch identifiers equivalent? Should I change one so they are both consistent?

charris replied:
The `upstream/master` form refers to remote branches that are available for merge. You can see them with a `git branch -a`. That form does not seem to work with `git pull`, I'm not sure why the two are different.

charris commented:
Looking good,I think this is almost ready, just a few more fixes.

rnelsonchem commented:
@charris @rgommers @juliantaylor 
I want to make sure that I'm going to make the proper changes to the "Main Repo" section:
Step 1: merge only. Even for single commits (Right now I have rebase as well, but from @charris comment it sounds like that leads to some accountability problems. 
Step 2: Do a git log to check things out
Step 3: git push git@https://github.com/numpy/numpy.git my-feature-branch:master

Does that cover everyone's concerns?

rnelsonchem commented:
Maybe it would be better to file a GitHub issue for a touch-up of the "Main Repo" section. Not being a core dev, I don't have a feeling for the best way to modify this section, and it sounds like there are a lot of differing opinions. 

Anyway, I fixed the other issues and squashed everything into one commit.

charris commented:
Just for my own future reference, we need to cover rebasing for backport, something like
```
git rebase --onto $(git merge-base master maintenance/1.9.x) HEAD^
```

charris commented:
@rnelsonchem Many thanks for updating this.
